src/passes/RemoveNonJSOps.cpp

```diff
@@ -14,251 +14,137 @@
  * limitations under the License.
  */
 
-//
 //
 // Removes all operations in a wasm module that aren't inherently implementable
-// in JS. This includes things like `f32.nearest` and
-// `f64.copysign`. Most operations are lowered to a call to an injected
+// in JS. This includes things like 64-bit division, `f32.nearest`,
+// `f64.copysign`, etc. Most operations are lowered to a call to an injected
 // intrinsic implementation. Intrinsics don't use themselves to implement
 // themselves.
 //
+// You'll find a large wast blob at the end of this module which contains all of
+// the injected intrinsics. We manually copy over any needed intrinsics from
+// this module into the module that we're optimizing after walking the current
+// module.
+//
 
 #include <wasm.h>
 #include <pass.h>
 
 #include "asmjs/shared-constants.h"
 #include "wasm-builder.h"
+#include "wasm-s-parser.h"
 
 namespace wasm {
 
+namespace {
+extern const char *IntrinsicsModule;
+}
+
 struct RemoveNonJSOpsPass : public WalkerPass<PostWalker<RemoveNonJSOpsPass>> {
-  bool needNearestF32 = false;
-  bool needNearestF64 = false;
-  bool needTruncF32 = false;
-  bool needTruncF64 = false;
-  bool needCtzInt32 = false;
-  bool needPopcntInt32 = false;
-  bool needRotLInt32 = false;
-  bool needRotRInt32 = false;
+  std::unique_ptr<Builder> builder;
+  std::unordered_set<Name> neededIntrinsics;
 
   bool isFunctionParallel() override { return false; }
 
   Pass* create() override { return new RemoveNonJSOpsPass; }
 
   void doWalkModule(Module* module) {
+    // Discover all of the intrinsics that we need to inject, lowering all
+    // operations to intrinsic calls while we're at it.
     if (!builder) builder = make_unique<Builder>(*module);
     PostWalker<RemoveNonJSOpsPass>::doWalkModule(module);
 
-    if (needNearestF32) {
-      module->addFunction(createNearest(f32));
-    }
-    if (needNearestF64) {
-      module->addFunction(createNearest(f64));
+    if (neededIntrinsics.size() == 0) {
+      return;
     }
-    if (needTruncF32) {
-      module->addFunction(createTrunc(f32));
-    }
-    if (needTruncF64) {
-      module->addFunction(createTrunc(f64));
-    }
-    if (needCtzInt32) {
-      module->addFunction(createCtz());
-    }
-    if (needPopcntInt32) {
-      module->addFunction(createPopcnt());
-    }
-    if (needRotLInt32) {
-      module->addFunction(createRot(RotLInt32));
-    }
-    if (needRotRInt32) {
-      module->addFunction(createRot(RotRInt32));
+
+    // Parse the wast blob we have at the end of this file.
+    //
+    // TODO: only do this once per invocation of wasm2asm
```

Comment:
why wouldn;t this be once per invocation of wasm2asm - it should run this pass only once?

Reply:
Currently the pass is run twice, before and after the i64 -> i32 lowering. The second time is due to the fact that the lowering pass can create intrinsics that weren't previously lowered. For example i64.ctz is lowered to i32.ctz, which is in turn lowered to an intrinsic.